pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -239,6 +248,56 @@ func (r *ReconcileUserAccount) ensureIdentity(logger logr.Logger, userAcc *toolc
 	return identity, false, nil
 }
 
+func (r *ReconcileUserAccount) ensureNSTemplateSet(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*toolchainv1alpha1.NSTemplateSet, bool, error) {
+	name := userAcc.Name
+
+	// create if not found
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name, Namespace: userAcc.Namespace}, nsTmplSet); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("creating a new NSTemplateSet", "name", name)
+			if err := r.setStatusProvisioning(userAcc); err != nil {
+				return nil, false, err
+			}
+			nsTmplSet = newNSTemplateSet(userAcc)
+			if err := controllerutil.SetControllerReference(userAcc, nsTmplSet, r.scheme); err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err,
+					"failed to set controller reference for NSTemplateSet '%s'", name)
+			}
+			if err := r.client.Create(context.TODO(), nsTmplSet); err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err,
+					"failed to create NSTemplateSet '%s'", name)
+			}
+			logger.Info("NSTemplateSet created successfully", "name", name)
+			return nsTmplSet, true, nil
+		}
+		return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err, "failed to get NSTemplateSet '%s'", name)
+	}
+
+	// update if not same
+	equal := compareNSTemplateSet(nsTmplSet.Spec, userAcc.Spec.NSTemplateSet)
+	if !equal {
+		nsTmplSet.Spec = userAcc.Spec.NSTemplateSet
+		if err := r.client.Update(context.TODO(), nsTmplSet); err != nil {
+			return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err, "failed to update NSTemplateSet '%s'", name)
+		}
+		return nsTmplSet, true, nil
+	}
+
+	// update status if ready=false
+	readyCond := condition.FindConditionByType(nsTmplSet.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	if readyCond != (toolchainv1alpha1.Condition{}) {
+		if readyCond.Status == corev1.ConditionFalse {
+			if err := r.setStatusFromNSTemplateSet(userAcc, readyCond.Reason, readyCond.Message); err != nil {
```

Comment:
It can be `False` with Reason `ProvisioningNamespace` and this is NOT a failed state rather in-process state.  So can't use `NsTemplateSetFailure` which may mis-guide that the state will be always failed.